src/nxt_otel.c

```diff
@@ -0,0 +1,403 @@
+        name = (nxt_str_t){
+            .start  = cur->name,
+            .length = cur->name_length,
+        };
+
+        val = (nxt_str_t){
+            .start  = cur->value,
+            .length = cur->value_length,
+        };
+
+        nxt_otel_rs_add_event_to_trace(r->otel->trace, &name, &val);
+    } nxt_list_loop;
+
+    method_name = (nxt_str_t){
+        .start  = (u_char *) NXT_OTEL_METHOD_TAG,
+        .length = nxt_length(NXT_OTEL_METHOD_TAG) + 1,
+    };
+    nxt_otel_rs_add_event_to_trace(r->otel->trace, &method_name, r->method);
+
+    path_name = (nxt_str_t){
+        .start  = (u_char *) NXT_OTEL_PATH_TAG,
+        .length = nxt_length(NXT_OTEL_PATH_TAG) + 1,
+    };
+    nxt_otel_rs_add_event_to_trace(r->otel->trace, &path_name, r->path);
+    nxt_otel_propagate_header(task, r);
+
+    nxt_otel_state_transition(r->otel, NXT_OTEL_BODY_STATE);
+}
+
+
+void
+nxt_otel_span_add_body(nxt_http_request_t *r)
+{
+    size_t    body_size, buf_size;
+    u_char    *body_buf, *body_size_buf;
+    nxt_str_t body_key, body_val;
+    nxt_int_t cur;
+
+    if (r->body != NULL) {
+        body_size = nxt_buf_used_size(r->body);
+    } else {
+        body_size = 0;
+    }
+
+    buf_size = 1; // first digit
+    if (body_size != 0) {
+        buf_size += log10(body_size); // subsequent digits
+    }
+    buf_size += 1; // \0
+    buf_size += nxt_length(NXT_OTEL_BODY_SIZE_TAG);
+    buf_size += 1; // \0
+
+    body_buf = nxt_mp_zalloc(r->mem_pool, buf_size);
+    if (body_buf == NULL) {
+        return;
+    }
+
+    cur = sprintf((char *) body_buf, "%lu", body_size);
+    if (cur < 0) {
+        return;
+    }
+
+    // (already was zero-alloc'ed)
+    //body_buf[cur] = '\0';
+    cur += 1;
+    body_size_buf = body_buf + cur;
+
+    nxt_cpystr(body_buf + cur, (const u_char *) NXT_OTEL_BODY_SIZE_TAG);
+    // (already was zero-alloc'ed)
+    //body_buf[cur] = '\0';
+
+    body_key = (nxt_str_t){
+        .start  = body_size_buf,
+        .length = nxt_length(body_size_buf),
+    };
+    body_val = (nxt_str_t){
+        .start  = body_buf,
+        .length = nxt_length(body_buf),
+    };
+
+    nxt_otel_rs_add_event_to_trace(r->otel->trace, &body_key, &body_val);
+    nxt_otel_state_transition(r->otel, NXT_OTEL_COLLECT_STATE);
+}
+
+
+void
+nxt_otel_span_add_status(nxt_task_t *task, nxt_http_request_t *r) {
+    u_char    *status_buf;
+    nxt_str_t status_key, status_val;
+
+    // dont bother logging an unset status
+    if (r->status == 0) {
+        return;
+    }
+
+    // add specific 3 character status to span
+    status_buf = nxt_mp_zalloc(r->mem_pool, 4);
+    if (status_buf == NULL) {
+        return;
+    }
+
+    sprintf((char *) status_buf, "%i", r->status);
+
+    // set up event
+    status_key = (nxt_str_t){
+        .start  = (u_char *) NXT_OTEL_STATUS_CODE_TAG,
+        .length = nxt_length(NXT_OTEL_STATUS_CODE_TAG),
+    };
+    status_val = (nxt_str_t){
+        .start  = status_buf,
+        .length = 4,
```

Comment:
If you remember the point of the nxt_str_t's from the last review, these are converted directly to Rust String types which I believe will need the \0 to be included.

Reply:
The string can have the terminating `\0`. We just don't normally include that in the length member. I.e. the length member would be strlen(buf).

Remember nxt_length() does sizeof() -1...

Reply:
I heard you the first time. As I said, the type conversion we are performing in the Rust code will need the length in this struct to account for the null terminator.

Reply:
Originally I had written this code to not use the nxt_str_t type, would you like me to revert it now that you know that we must use it slightly differently here?

Reply:
Well you _don't_ account for it in status_key...

If you _really_ think you need the `\0` accounted for in the `.length` member then please do so on the rust side.

Assuming

```c
    status_key = (nxt_str_t){                                                   
        .start  = (u_char *) NXT_OTEL_STATUS_CODE_TAG,                          
        .length = nxt_length(NXT_OTEL_STATUS_CODE_TAG),                         
    };
```

becomes

```c
nxt_str_set(&status_key, NXT_OTEL_STATUS_CODE_TAG);
```

Then you'd need to do that on the rust side anyway.

But please show me where it's required in the rust code.

Looking at the rust code I'm not convinced you do, otherwise you already have a problem...

```rust
        let key = String::from_utf8_unchecked(                                  
            slice::from_raw_parts((*key).start, (*key).len).to_vec()            
        ).clone();                                                              
        let val = String::from_utf8_unchecked(                                  
            slice::from_raw_parts((*val).start, (*val).len).to_vec()            
        ).clone();
```

`key.len` does _not_ account for the `\0`.

AIUI rust `String`'s are _not_ inherently nul-terminated.

As it stands `key` is _not_ null terminated. `val` is. If these are only now being used natively in rust then neither should need to be, right?